rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,435 @@
+
+Finally distinguish between doc-strings and comments.
+
+# Motivation
+[motivation]: #motivation
+
+This RFC aims at improving consistency of in code documentation. (aka Doc-strings)
+
+The community offers tools and methods (such as nixdoc, nix-doc, etc.) to write and process in-code documentation for functions and other code related atomic expressions.
+This functionality is currently utilized to build a subset of documentation for nix functions. (e.g. nixpkgs.lib documentation via: [nixdoc](https://github.com/nix-community/nixdoc))
+There is also [noogle](https://noogle.dev) that indexes subsets of nixpkgs based on the current comments.
+
+However the format of that __docs-strings__ is itself neither well documented nor standardized.
+
+This RFC aims at achieving consistency for docs-strings and allow to differentiate between regular comments and doc-strings.
+
+We could envision native nix support, documentation-team or community-driven solutions for automatically generating documentation from them.
+
+__Current issue__
+
+> The existing doc-strings heavily depend on a tool called `nixdoc` and not vice versa.
+>
+> Instead we want to provide a common standard that every nix user can refer to.
+
+Everything until now is just a draft if you can provide better ideas e.g. using different formats or syntax please let me know.
+
+## Example of the current format:
+
+```nix
+/*
+  <Description
+  Example:
+  <Code Examples>
+  Type:
+  <Some Type information>
+*/
+expr = 
+# Describes 'arg'
+arg:
+# Describes 'foo'
+foo: 
+```
+
+> This RFC aims for general rules for doc-strings.
+> Features like: "what different sections exist" and if the might have complex rules (e.g. type: syntax) is not specified.
+>
+> Providing a formal skeleton, where sections can be extended by the nix community
+
+## Proposed solution
+
+```nix
+# somefile.nix
+
+    ## <Description or Tagline>
+    ## 
+    ## # Example
+    ##
+    ## <Comprehensive code>
+    ## 
+    ## # Type
+    ##    
+    ## <Type Signature>
+    mapAttrs = f: s: #...
+```
+
+This is considered a possible solution because the following reasons:
+
+- Good visual distinction between doc-strings and comments
+- Clear rules and structure
+- Saves vertical space
+- Doesn't need termination (*/)
+- Clear indentation rules. Following the __markdown__ convention.
+
+# Detailed design
+[design]: #detailed-design
+
+The following abstract rules describe how to write doc-strings.
+
+> I'am very happy if you comment about whether we should use `## {content}` or `/** {content} */`
+> I did write this RFC tailored towards `##` but using `/** */` is still an open discussion.
+
+We must find one solution out of the following:
+
+|  | 0 `##` | 1 `/** */`   |
+|---|---|---|
+| 0 `# {Keyword}` | `## # Example` | `/** # Example */` |
+| 1 `@{Keyword}:`  | `# @Example:` |`/** @Example: */` |
+
+Proposed Solution (0,0) => `##` For docstring body and markdown headings `# H1`
+
+### Format Rules
```

Comment:
maybe we should limit doc-strings to annotate only attribute names in attribute set and let-in bindings?

Reply:
Why would you only want to document sets? 
They are the most prominent points that need to be documented. But this also restricts the usage quite a lot. 
There might be less-known places that need docs for anonymous functions or other things that we didn't yet think of.

Reply:
Otherwise we would need to define associativity for doc-strings, so we can know whether `docs` annotates `foo` or `foo bar baz`
```nix
## docs
foo bar baz
```

In rust, doc comments can only document things that are named. AFAIK javadoc has the same limitation

I didn't mention lambda parameters, but annotating those also sounds good to me. Can you give an example where documenting an anonymous function or things that are not attribute names would be helpful?

Reply:
I am not sure. 

https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix

I always wondered how to properly document mkDerviation.

Reply:
good point, maybe we can make doc-strings have the lowest precedence when it comes to associativity, or just not allow it when it can be ambiguous, which will allow documenting anonymous functions

Reply:
I think i‘ll add your exact example and say the precedence is always lowest. (E.g. -1) 
So this would actually be a docstring for „foo bar baz“